_posts/2020-11-12-Typescript-in-a-serverless-backend-with-NestJS-and-Azure-Functions.md

```diff
@@ -0,0 +1,317 @@
+
+The user can send new data to the application, which will keep the history of the three parameters in a CosmosDB.
+Apart from the main `AppModule`, I only added two modules, the `ParametersModule` and the `LoggerModule` to the application.
+This blog post won't explain building the `LoggerModule`, as it only serves to create a custom logger.
+The [NestJS documentation](https://docs.nestjs.com/techniques/logger){:target="_blank" rel="noopener noreferrer"} provides a very clear explanation of how to use a custom logger.
+
+The application has two (basic) guards set up for the HTTP calls, one for authorization and one for role-based access. 
+For this guard, we use a simple bearer token, however, integration with eg. Cognito or JWT tokens is available.
+
+## ParametersModule
+
+A Module in NestJS provides a clear way of organizing the project and enabling clear dependency injection.
+There are four important things you can mark within a module: 
+- `controllers`: classes which capture incoming HTTP calls
+- `providers`: classes marked with NestJS's `@Injectable()`, made available for dependency injection
+- `imports`: modules that need to be imported, again for dependency injection
+- `exports`: subset of the providers that need to be exported for use in other modules
+
+The `ParametersModule` imports three other modules: the previously explained `LoggerModule`, the `ConfigModule` and the `AzureCosmosDBModule`.
+The `ConfigModule` is used to be able to access environment variables from a `.env` file or from the configuration of the Azure Function. 
+Note that this is also possible with a package such as `dotenv`, however this isn't very ideal as we would have to access `process.env` directly every time.
+Of course, pushing those environment files to git is bad practice. 
+You can find a `.env-sample` file in the repository, which is used to show which variables need to be filled in the `.env` file.
+
+```js
+@Module({
+  controllers: [ParametersController],
+  providers: [
+      ParametersService, 
+      ParametersRepository,
+      {
+        provide: APP_GUARD,
+        useClass: RolesGuard,
+      },
+  ],
+  imports: [
+      LoggerModule,
+      ConfigModule,
+      AzureCosmosDbModule.forFeature([{dto: ParametersEntity}])
+  ]
+})
+export class ParametersModule {}
+```
+
+The first step when receiving an HTTP request to the application is the `ParametersController` as shown below.
+This controller will catch all requests on the 'parameters' endpoint.
+Using annotations, you can:
+- make a check for the type of incoming request and divide the traffic accordingly. This is similar to the way annotations work in the spring framework (eg. `@PostMapping()`).
+- customise which Http code you want to return on successful calls, as I did with the `createParameters` method.
```

Comment:
I'd write it as HTTP to be consistent with the other usages throughout the blogpost.
```suggestion
- Customise which HTTP code you want to return on successful calls, as I did with the `createParameters` method.
```